sphericart/src/cuda_base.cpp

```diff
@@ -0,0 +1,257 @@
+    dim3 grid_dim(find_num_blocks(nedges, GRID_DIM_Y));
+
+    size_t smem_size =
+        total_buffer_size(l_max, GRID_DIM_X, GRID_DIM_Y, sizeof(scalar_t), gradients, hessian);
+
+    int _nprefactors = nprefactors;
+    int _lmax = l_max;
+    int _n_total = n_total;
+    int _nedges = nedges;
+    bool _gradients = gradients;
+    bool _hessian = hessian;
+    bool _normalize = normalize;
+
+    std::vector<void*> args = {
+        &xyz,
+        &_nedges,
+        &prefactors,
+        &_nprefactors,
+        &_lmax,
+        &_n_total,
+        &_gradients,
+        &_hessian,
+        &_normalize,
+        &sph,
+        &dsph,
+        &ddsph
+    };
+
+    std::string kernel_name = getKernelName<scalar_t>("spherical_harmonics_kernel");
+    auto& kernel_factory = KernelFactory::instance();
+
+    CachedKernel* kernel = kernel_factory.create(
+        kernel_name,
+        std::string(CUDA_CODE),
+        "sphericart_impl.cu",
+        {"--define-macro=CUDA_DEVICE_PREFIX=__device__"}
+    );
+
+    kernel->launch(grid_dim, block_dim, smem_size, cstream, args);
+}
+
+template void sphericart::cuda::spherical_harmonics_cuda_base<float>(
+    float* xyz,
+    const int nedges,
+    float* prefactors,
+    const int nprefactors,
+    const int64_t l_max,
+    const bool normalize,
+    const int64_t GRID_DIM_X,
+    const int64_t GRID_DIM_Y,
+    const bool gradients,
+    const bool hessian,
+    float* sph,
+    float* dsph,
+    float* ddsph,
+    void* cuda_stream
+);
+
+template void sphericart::cuda::spherical_harmonics_cuda_base<double>(
+    double* xyz,
+    const int nedges,
+    double* prefactors,
+    const int nprefactors,
+    const int64_t l_max,
+    const bool normalize,
+    const int64_t GRID_DIM_X,
+    const int64_t GRID_DIM_Y,
+    const bool gradients,
+    const bool hessian,
+    double* sph,
+    double* dsph,
+    double* ddsph,
+    void* cuda_stream
+);
+
+template <typename scalar_t>
+void sphericart::cuda::spherical_harmonics_backward_cuda_base(
+    scalar_t* dsph,
+    scalar_t* sph_grad,
+    const int nedges,
+    const int ntotal,
+    scalar_t* xyz_grad,
+    void* cuda_stream
+) {
+
+    static const char* CUDA_CODE =
+#include "generated/wrapped_sphericart_impl.cuh"
+        ;
+    std::string kernel_name = getKernelName<scalar_t>("backward_kernel");
+
+    auto& kernel_factory = KernelFactory::instance();
+
+    dim3 block_dim(4, 32);
+    auto find_num_blocks = [](int x, int bdim) { return (x + bdim - 1) / bdim; };
+    dim3 grid_dim(find_num_blocks(nedges, 32), 3);
+    cudaStream_t cstream = reinterpret_cast<cudaStream_t>(cuda_stream);
+
+    int _n_total = ntotal;
+    int _nedges = nedges;
+
+    std::vector<void*> args = {&dsph, &sph_grad, &_nedges, &_n_total, &xyz_grad};
+
+    CachedKernel* kernel = kernel_factory.create(
+        kernel_name,
+        std::string(CUDA_CODE),
+        "sphericart_impl.cu",
+        {"--define-macro=CUDA_DEVICE_PREFIX=__device__"}
```

Comment:
why can't this be hardcoded inside the `.cu` file? Do we ever define it to something different?

Reply:
it could be hardcoded there. I'll give it a try.

Reply:
should be resolved in latest commit.